File: ruby/33.search-in-rotated-sorted-array.rb
# @param {Integer[]} nums
# @param {Integer} target
# @return {Integer}
def search(nums, target)
  left = 0
  right = nums.length - 1

  while left <= right
    mid = (left + right) / 2
    return mid if nums[mid] == target

    if nums[left] <= nums[mid]
      if nums[left] <= target && target < nums[mid]
        right = mid
      else
        left = mid + 1
      end
    elsif nums[mid] < target && target <= nums[right]
      left = mid + 1
    else
      right = mid
    end
  end

  -1
end

File: go/code/33.search-in-rotated-sorted-array.go
package main

func search(nums []int, target int) int {
	left := 0
	right := len(nums) - 1

	if left == right {
		if nums[left] == target {
			return 0
		} else {
			return -1
		}
	}

	for left <= right {
		mid := left + (right-left)/2

		if nums[mid] == target {
			return mid
		}

		// border is nothing
		if nums[left] < nums[mid] && nums[mid] < nums[right] {
			if target < nums[mid] {
				right = mid
			} else {
				left = mid + 1
			}
		} else {
			// border is on the right
			if nums[left] <= nums[mid] {
				if nums[left] <= target && target < nums[mid] {
					right = mid
				} else {
					left = mid + 1
				}
			} else {
				// border is on the left
				if nums[mid] < target && target <= nums[right] {
					left = mid + 1
				} else {
					right = mid
				}
			}
		}
	}

	return -1
}
